crates/provider/src/ext/trace.rs

```diff
@@ -0,0 +1,117 @@
+//! This module extends the Ethereum JSON-RPC provider with the Trace namespace's RPC methods.
+use crate::{Provider, RpcWithBlock};
+use alloy_eips::BlockNumberOrTag;
+use alloy_network::Network;
+use alloy_primitives::TxHash;
+use alloy_rpc_types_trace::parity::{LocalizedTransactionTrace, TraceResults, TraceType};
+use alloy_transport::{Transport, TransportResult};
+
+/// List of trace calls for use with [`TraceApi::trace_call_many`]
+pub type TraceCallList<'a, N> = &'a [(<N as Network>::TransactionRequest, Vec<TraceType>)];
+
+/// Trace namespace rpc interface that gives access to several non-standard RPC methods.
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+pub trait TraceApi<N, T>: Send + Sync
+where
+    N: Network,
+    T: Transport + Clone,
+{
+    /// Executes the given transaction and returns a number of possible traces.
+    ///
+    /// # Note
+    ///
+    /// Not all nodes support this call.
+    fn trace_call<'a, 'b>(
+        &self,
+        request: &'a N::TransactionRequest,
+        trace_type: &'b [TraceType],
```

Comment:
cc @mattsse my guess is we could also return a builder that's called `TraceRpcWithBlock` or similar? since we default to `["trace"]` most times?